P145 gases   m qxxx (mq135 co2, mq3 alcohol)

flashmark opened the pull request:
Adding a plugin for simple gas sensors based upon a heated electrode. Input is an analog value measured by an ESP analog input. These sensors are sold for low prices as a series of MQ-xxx. Each sensor type is tuned to a specific set of gases. Drawback is they are not very specific and require tuning. This plugin supports a subset of the MQ-xxx sensors and uses a table with the sensor specific parameters. Plugin has been tested with an MQ-135 CO2 sensor. It is seen as "under development". See comment at the top of the plugin file.

flashmark committed:
Introduce P145 Gasses MQ-xxx

flashmark committed:
P145 cleanup, added sensors, start ESP32 support

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
P145 Cleanup, update calibrated value

flashmark committed:
Merge branch 'P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)' of https://github.com/flashmark/ESPEasy into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
P145 add [TESTING]

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,282 @@
+#define PCONFIG_TEMP_VAL    PCONFIG(2)
+#define PCONFIG_HUM_TASK    PCONFIG(3)
+#define PCONFIG_HUM_VAL     PCONFIG(4)
+#define PCONFIG_SENSORT     PCONFIG(5)
```
Please prefix these defines with the plugin NR.
These are rather generic names and when these would be used elsewhere in the code you would get all kinds of freaky issues.

For example:
`#define PCONFIG_SENSORT PCONFIG(5)` -> `#define P145_PCONFIG_SENSORT PCONFIG(5)`
And of course for all others in this block too.

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,282 @@
+      }
+      for (int i=0; i<x; i++)
+      {
+        options[i] = (String)P145_data_struct::getTypeName(i) + (String)(F(" - ")) + (String)P145_data_struct::getGasName(i);
```
Please use the `concat` function here:
```c++
options[i] = concat(P145_data_struct::getTypeName(i), F(" - ")) + P145_data_struct::getGasName(i);
```

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,282 @@
+      addADC_PinSelect(AdcPinSelectPurpose::ADC_Touch_HallEffect, F("taskdevicepin1"), CONFIG_PIN1);
+# endif // ifdef ESP32
+
+      addFormTextBox(F("Load Resistance"), F("plugin_145_RLOAD"), String(PCONFIG_FLOAT_RLOAD), 33);
```
Why not using a float type box?

flashmark replied:
Still learning...

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,282 @@
+      PCONFIG_HUM_VAL   = getFormItemInt(F("plugin_145_humidity_value"));
+
+      P145_data_struct *P145_data = new (std::nothrow) P145_data_struct();
+      initPluginTaskData(event->TaskIndex, P145_data);
```
Do not init the plugin on saving.
When a task is saved, the `PLUGIN_EXIT` will be called, followed by `PLUGIN_INIT` (when enabled)
So there is no need to init anything when saving it.
If there is something special where `PLUGIN_EXIT` should be called before `PLUGIN_SAVE` or after, then you can set `Device[deviceCount].ExitTaskBeforeSave = false;` (or `true`) in the `PLUGIN_DEVICE_ADD` call.

flashmark replied:
Thanks, copy-paste error. Intended was to access the current P145_data_struct

TD-er commented on src/src/PluginStructs/P145_data_struct.h:
```diff
@@ -0,0 +1,104 @@
+  float core;         // CORE
+  float corf;         // CORF
+  float corg;         // CORG
+};
```
Please sort the members of this class with the "largest" types first.
`float` is 4 byte-aligned, char-array is 1 byte aligned and I have no clue what the enum will be.
That's the real problem, that the alignment will be compiler dependent and thus you might get really funky issues when such struct data is stored as PROGMEM.
You can force a type on the enum, like this: `enum P145_SENSORDEF : uint32_t`
I know this is way too large for what is needed, but this way you will have a multiple of 4 bytes as struct size, which will probably cause no issues when reading it from PROGMEM.

Anyway, it is always a good idea to sort members based on their size as it will lead to the smallest possible structs in memory.

flashmark replied:
Real embedded thinking :-) I really like such optimizations when required, as software architect I prefer a "logical" schema, especially as others will append the table with sensor types. Strings were deliberately set 8 charecters (or 16 if 8 would not fit). I will rearrane this as the table could become big. Although there will be one copy in ram. I don't know how PROGMEM data works. If possible the P145_data_struct code could access this static data through a pointer directly in PROGMEM. For now I will rearrange it.

TD-er replied:
PROGMEM data is read per 32 bits at once, so better to have it all in multiples of 4 bytes.
Alignment may be an issue with PROGMEM stored structs. Therefore I rather have the structs sized at multiples of 4 bytes.
If you don't sort the members based on size, you may end up with either a larger struct, or the compiler may decide to make it a packed struct. Then the typically 4-byte types may no longer be 4-byte aligned and then you get all kinds of nasty things.
So better take extra care when using PROGMEM storage. Sometimes you need to read it first into regular memory using those `_P` postfixed functions and then map it to a struct or something.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+      0.0,          // CORG   
+  }
+};
+const int nbrOfTypes = (int)(sizeof(sensorData) / sizeof(struct P145_SENSORDEF));
```
Please mark this as a `constexpr`.
When marked as such, the compiler will compute it at compile time and simply use the result in the code instead of actually computing it at run-time.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+void P145_data_struct::calibrate (float currentRcal)
+{
+  unsigned long now = millis();
+  unsigned long time = now - last_cal;
```
Please use `timePassedSince`.
This way we know for sure it will be the correct time calculation and not cause a typical "49.7 days bug"

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+  {
+    rcal = currentRcal;
+  }
+  doit = (time > P145_CALIBRATION_INTERVAL) && (rcal > 0.0);
```
This can be a `const bool` as it is not changed after setting it.

flashmark replied:
I would expect the compiler to optimize it. From a logical point of view one can argue if it is really intended as const. But I will change it as it is indeed intended to reuse exactly teh same condition twice.

TD-er replied:
Setting things to const is also to prevent bugs.
For example you want to check `==` by due to a typo you type `=`.
When you force yourself to have the const type on the left, you will be prompted by the compiler (or even the intellisense highligher) you made an error.

Just make functions and variables const whenever you can and if the compiler starts complaining you need to double check whether it was an oversight to make it const. But most of the time it is just a bug which is detected before anyone else saw it (apart from the compiler :) )

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+        VCC           Voltage applied to the sensor-Rload combination
+*/
+/*****************************************************************************/
+float P145_data_struct::getResistance(float val)
```
This can be a const function.
```c++
float P145_data_struct::getResistance(float val) const
```
(don't forget to also add `const` in the .h file)

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+  switch (algorithm)
+  {
+    case p145AlgA:  // MQ-135
+      newValue = (rSensor * pow((refLevel / sensordef.para), (1.0 / sensordef.parb)));
```
please use `powf` here as the compiler would otherwise convert the floats to `double` and use the `double` version of `pow`.

Also every "magic value" of a float should be written with a `f`.
Thus `1.0` -> `1.0f`, for the same reason as to use `powf` when using floats.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+      }
+      break;    
+  };
+  return (newValue);
```
No need to wrap the `newValue` in braces here.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+  {
+    c = sensordef.core * temperature + sensordef.corf * humidity + sensordef.corg;
+  }
+  return (rSensor / c) * pow((refLevel / sensordef.para), (1. / sensordef.parb));
```
same here regarding the trailing `f` for float values and use of `powf`

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+@return The corrected sensor resistance RZero in kOhm
+*/
+/*****************************************************************************/
+float P145_data_struct::getCorrectedRZero(float rSensor, float temperature, float humidity)
```
Also a `const` function.

flashmark replied:
Compiler does't like it to be const

TD-er replied:
In that function you call `getRZero`.
As long as that function isn't `const`, all functions calling it cannot be `const`.
But as far as I can see, both functions do not change members.
So unless it is marked as a `static` function, then both should be able to set as `const`.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+      return (0.0);
+      break;
+  }
+  return 0.0;   // Default value should never be returned
```
Same as before (traling `f`, const function)

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+      return (0.0);
+      break;
+  }
+  return 0.0;  // Default value should never be returned
```
dito.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+      break;
+    default:
+      return (0.0);
+      break;
```
No need for `break;` after a return.
And since you also return the same value at the end, you can leave out the return statement for the default case anyway.

flashmark replied:
Defensive programming. Juggling between two thoughts 1) Return the intended velue for each case 2) Safety net at the end. As far as I know the compiler complains when not having a return at the end. I will clean it up a bit.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+        sensor type.
+*/
+/**************************************************************************/
+float P145_data_struct::getCalibrationValue()
```
`const` function

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+@note   This is the internally corrected Rzero and will be unaltered when
+        Autocalibration is switched off.
+*/
+float P145_data_struct::getAutoCalibrationValue()
```
`const` function

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+@note   This function is only used for debugging
+*/
+/**************************************************************************/
+void P145_data_struct::dump()
```
`const` function

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+    String log;                       // Helper string to build log text
+    log = F("MQ-xx: NAME ");
+    log += sensordef.name;
+    addLog(LOG_LEVEL_INFO, log);
```
Why not like this:
```c++
    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: NAME "), sensordef.name));
```
This way there is no need for allocating a `log` variable, it makes the code more readable (in this use case) and the return value of `concat` will be moved using `std::move` instead of being copied.

flashmark replied:
Sounds good to me, I was just copying existing patterns...

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+         This is a static member function
+*/
+/**************************************************************************/
+const __FlashStringHelper * P145_data_struct::getTypeName(int stype)
```
`const` function

flashmark replied:
Does not match with static qualifier

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+         This is a static member function
+*/
+/**************************************************************************/
+const __FlashStringHelper * P145_data_struct::getGasName(int stype)
```
`const` function

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,655 @@
+   @note This is a static member function
+*/
+/**************************************************************************/
+int   P145_data_struct::getNbrOfTypes()
```
`const` function

TD-er commented:
OK, I had a quick glance at your code.
It may not look like it, but I really do like the neat and structured coding style!

tonhuisman commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,282 @@
+      addFormCheckBox(F("Enable temp/humid compensation"), F("plugin_145_enable_compensation"), compensate);
+      addFormNote(F("If this is enabled, the Temperature and Humidity values below need to be configured."));
+      // temperature
+      addHtml(F("<TR><TD>Temperature:<TD>"));
+      addTaskSelect(F("plugin_145_temperature_task"), PCONFIG_TEMP_TASK);
```
This `addHtml(...);` can be replaced by `addRowLabel(F("Temperature"));`, thus using existing functions and shorter strings.

and the variable names like `"plugin_145_temperature_task"` can do without the `plugin_145_` prefix (there will be only 1 page on screen at any moment), and preferably all lower-case (above), also saving a few bytes in .bin size (have started a PR cleaning/compacting all those unneeded prefixes and other not really needed long strings for .bin size reason...)

flashmark committed:
Rework during review

flashmark commented:
> OK, I had a quick glance at your code.
It may not look like it, but I really do like the neat and structured coding style! 

I did a lot of embedded system coding in C, but now work as an embedded software architect with Microsoft office as main tools :-( 
Experience in C++ is very limited. Any help is useful. And I am responsible for code that will live over 30 years.......

TD-er commented:
> And I am responsible for code that will live over 30 years.......

Ah, so you're the one to eventually fix the 19 January 2038 bug ;)

flashmark commented:
I see one build failing on memory size. Strange as I did not add the P145 to any set yet. I also see complains about branch protection. What actions are expected from me as contributor to get the code merged?
Thanks for assisting me in advance, as you can see I am not very familiar with the Github processes.

TD-er commented:
You can edit these two files and remove 2 plugins from it:

https://github.com/letscontrolit/ESPEasy/blob/mega/tools/pio/pre_custom_esp82xx.py
https://github.com/letscontrolit/ESPEasy/blob/mega/tools/pio/pre_custom_esp32.py

For now, remove these 2 lines from both files:
```
    "-DUSES_P085",  # AcuDC24x
    "-DUSES_P100",  # Pulse Counter - DS2423
```

And then push that commit.
I will then approve the new build on this PR.

I can also do it, but I am not sure how familiar you are with Git.

flashmark committed:
Removed P085 & P100 from test setup to free some space
P145 Tidy up some comments

flashmark commented:
I changed those settings in my own branch (Visual studio code) , I assume that was expected. Pushed the changes tio Github.

TD-er commented:
> I changed those settings in my own branch (Visual studio code) , I assume that was expected. Pushed the changes tio Github.

Yep, that's what I meant indeed.
Just approved to run the test build :)

tonhuisman commented on tools/pio/pre_custom_esp32.py:
```diff
@@ -42,10 +42,10 @@
     "-DUSES_P098",  # PWM motor
 
-    "-DUSES_P100",  # Pulse Counter - DS2423
+#    "-DUSES_P100",  # Pulse Counter - DS2423
```
For ESP32 these don't have to be disabled, they will still fit in the build

flashmark committed:
Merge branch 'mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
Rollback removal of P085 and P100 exclusion in prwe_custom_esp32.py

flashmark committed:
P145 start documentation

flashmark commented:
I started documentation. I wonder if it is possible to pull in this plugin to mega already? Probabaly for testing without adding it to any set already.

TD-er committed:
Merge branch 'mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

TD-er commented:
> 

I just merged 'mega' into this PR and resolved your merge conflict.
Let me have a look once more at the code to see if there is something obviously 'wrong' and if not, then I see no reason why it can't already be merged.

tonhuisman commented:
The documentation links are not in the `_Plugin.rst` and `_plugin_categories.rst` pages yet.

TD-er commented on src/src/PluginStructs/P145_data_struct.h:
```diff
@@ -0,0 +1,106 @@
+  float corg;         // CORG
+  P145_algorithm alg; // Preferred/tuned algorithm
+  char name[8];       // Sensor type name
+  char gas[8];        // Measured gas concentration
```
No member is initialized during construction.
The char arrays can be initialized like this:
`char gas[8] = {0};`
The floats can be initialized like this:
`float para = 0.0f;`
or the more generic and new (since C++11) approach:
`float para{};` or `float para = {};`   which will also initialize it to the most sensible default value for these basic types which do not have their own constructor. (0.0f for floats)
N.B. for types which do have a constructor, calling `{}` or `={}` will actually call their constructor.
Usually for just members of some struct or class type, their constructor is being called. However for arrays of such a type it is not guaranteed their constructor is called.  Not sure if this is also behavior as is defined in the standard, but I've experienced this recently where members of some array were not initialized.

So better to always initialize explicitly. Either by assigning some value, or adding the `{}` when declaring them in a .h file.

flashmark replied:
I fully agree with proper initialization. In this case the structure is explicitly defined to hold a table of predefined sensor types in PROGMEM. There is one copy (you may call it a cache) in P145_data_struct. This will pretty well secure integrity. However let's keep to the project's coding style.

flashmark replied:
I need some help here as I am not a C++ expert. When I initialize teh fiels teh compiler complains about the hard coded "const struct P145_SENSORDEF sensorDefs[] PROGMEM" definition in P145_data_struct.cpp. I need a table with many sensors with each sensors characteristics and that data should be kept in PROGMEM. Only for the actual sensor the data is copied to RAM.

TD-er replied:
OK, please help me remind to check this tomorrow morning...

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+@note   The sensor is expected to be sampled ten times per second.
+        Lowest and highest value is dropped to remove outliers.
+        All other values are averaged to a single float value.
+        This algorithm is copied from P002_ADC.
```
I think we should turn this into some statistics class, as this code gets duplicated in more places.
The same for an IIR or FIR filter.

flashmark replied:
I copied it from the analog input as it is a good and relatively cheap algorithm to filter the analog input signal over the full measurement period. Dropping outliers seems also be handy on the ESP. As a beginning contributer without deep knowledge of C++ I will wait for others to pick it up. For this sensor it is very helpful as the active signal range is pretty small.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+// This table is stored in PROGMEM to save space in RAM
+// The data for the selected sensor type is copied to RAM, see P145_data_struct
+/******************************************************************************/
+const struct P145_SENSORDEF sensorData[] PROGMEM =
```
Maybe the naming of this array isn't correct.
sensorData suggests "runtime" or "live" data.
But it actually is a sensor definition or some kind of template.
But given its type is called SENSORDEF, we should call the array `sensorDefs`

TD-er commented on src/src/PluginStructs/P145_data_struct.h:
```diff
@@ -0,0 +1,106 @@
+    /* Sensor type & user options */
+    bool compensation = false;            // Use temperature compensation
+    bool calibration = false;             // Perform auto calibration
+    int  sensorType = 0;                  // Selected sensor type
```
This value is now initialized by 0, but I guess it should be set to -1.
If you call the `setSensorData` function and intend to load the values for type 0, it will not load it.

flashmark replied:
A very good catch

TD-er commented on src/src/PluginStructs/P145_data_struct.h:
```diff
@@ -0,0 +1,106 @@
+    bool compensation = false;            // Use temperature compensation
+    bool calibration = false;             // Perform auto calibration
+    int  sensorType = 0;                  // Selected sensor type
+    P145_algorithm algorithm = p145AlgA;  // conversion algorithm
```
This is also an incorrect initialization.
We would assume sensortype 0 is active, which has p145AlgNone als algorithm, not this one.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+    float ain = getAnalogValue();           // Analog measured and filtered data 
+    float rSensor = getResistance(ain);     // Convert to sensor resistance Rs
+    float rCal    = 0.0;                    // Potential Rzero for calibration
+    float value = 0.0;                      // Return value
```
`0.0` instead of `0.0f`
This means the compiler will make a double type first and then casts it to a float.
There are several other places in this code where the `f` is missing.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+    }
+
+    // Perform calibration if functionality is enabled
+    if (calibration && (rCal < 1.0e6)) 
```
also `1.0e6` is not of type `float`

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+    {
+      addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: level= "), value));         // Calculated sensor value
+#ifdef P145_DEBUG
+      addLog(LOG_LEVEL_INFO, concat(concat(F("MQ-xx: Sensor type= "), sensorType), concat(F(": "), sensordef.name))); 
```
Please wrap the xxx.name and xxx.gas types into something like this:  `String(sensordef.name)` when used along with `concat`
The reason is that `concat` is a template function.
And since `name` is of type `char[8]`, the compiler will create a new instance of each call like this.
Right now `gas` and `name` are of the same length, but when calling concat with a `char[9]` it will generate new code for the same function.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+
+/*****************************************************************************/
+/*!
+@brief  Handle the 10Hz processing for the plugin strct/class 
```
typo in the comment :)

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+const __FlashStringHelper * P145_data_struct::getGasName(int stype)
+{
+  if ((stype < nbrOfTypes) && (stype >= 0))
+    return (const __FlashStringHelper *)&sensorData[stype].gas;
```
This really is tricky.
Either the type in that struct should then be changed into `const __FlashStringHelper *` instead of `char[8]` or just return type `String`  for this function.
I think casting a char array to a flash string is extremely tricky as there is no guarantee how other functions may try to access this data.
What if there is some check somewhere in the SDK code to check if the address is in the expected range.
No, please change this as it will for sure cause issues and I have no idea how bad those will be.

flashmark replied:
Yeah, I found this somewhere when I was in doubt about how to treat a char[] in PROGMEM. I start to like the magic of the String class. But as an old school embedded C programmer I always tried to keep the number of constructors/destructors in expressions low. But returning a String object makes it much more safe and readable.

TD-er replied:
You can return of type flashstring, but then the members in the PROGMEM struct should also be of this type.
You can interpret flash strings as char pointers and access them using the standard functions with the `_P` postfix (e.g. `memcpy_P`) but not the other way around.
You cannot cast something which isn't a flash string into a flashstring type.
Well you can and the compiler doesn't think it's wrong, but you don't know how other functions will try to access the content of something that's marked as a flash string.

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+  {
+    c = sensordef.core * temperature + sensordef.corf * humidity + sensordef.corg;
+  }
+  return (rSensor / c) * pow((refLevel / sensordef.para), (1.0f / sensordef.parb));
```
Please use `powf`

TD-er commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -0,0 +1,598 @@
+/*****************************************************************************/
+float P145_data_struct::getResistance(float val) const
+{
+  return ((MAX_ADC_VALUE * P145_VCC) / (P145_VMAX * val) - 1.0f) * rload;
```
What if `val` equals 0?

flashmark replied:
What I have seen is that the formula is executed with a result of "inf". No crashes, but value out of range. It is debatable if we should rely on this or proactive check for boundaries. Problem is what happens if it is very close to zero...
I will add boundary checking at the cost of some extra code.

TD-er replied:
See the functions in `src/Helpers/ESPEasy_math.h` :)

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,281 @@
+      addFormFloatNumberBox(F("R Zero"), F("plugin_145_RZERO"), P145_PCONFIG_RZERO, 0.0, 10e6, 2U);
+      addUnit(F("Ohm"));
+       addFormFloatNumberBox(F("Reference Level"), F("plugin_145_REFLEVEL"), P145_PCONFIG_REF, 0.0, 10e6, 2U);
+      addUnit(F("ppm"));
```
Lots of occurences of `double` type values.

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,281 @@
+      }
+      for (int i=0; i<x; i++)
+      {
+        options[i] = concat(P145_data_struct::getTypeName(i), F(" - ")) + (String)P145_data_struct::getGasName(i);
```
Do not use C-style casts like this. Especially not for type `String`
A C-style cast is a bit like a `reinterpret_cast` which essentially tells the compiler "interpret the data starting at that address as if it were initialized/constructed as the type I tell you it is".
However for a `String` class, this just isn't correct.
The `String` class does have some buffer memory and some flags on how to interpret this buffer. It either is upto 12 bytes of a C-style char array, or a pointer. But if you try to use this `String` object, you have to hope the functions trying to do something with its data isn't concluding this text data may need to be intepreted as a pointer since the bits telling so are undefined.

Simple fix here is to change `(String)P145_data_struct::getGasName(i)` into `String(P145_data_struct::getGasName(i))`.

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,281 @@
+        float calVal = P145_data->getCalibrationValue();
+        if (calVal > 0.0) 
+        {
+          addFormNote((String)F("Current measurement suggests Rzero= ") + (String)calVal);
```
Another C-style cast to a String.

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,281 @@
+      addHtml(F("<TR><TD>Temperature Value:<TD>"));
+      addTaskValueSelect(F("plugin_145_temperature_value"), P145_PCONFIG_TEMP_VAL, P145_PCONFIG_TEMP_TASK);
+      // humidity
+      addHtml(F("<TR><TD>Humidity:<TD>"));
```
See markup functions like `html_TR_TD()` , `html_TR()`, etc.

And for this special case you do here, we have the function `addRowLabel`

tonhuisman replied:
That could even be an `addRowLabel(F("<Label without a colon>"))`

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,281 @@
+      addFormNote(F("If this is enabled, the Temperature and Humidity values below need to be configured."));
+      // temperature
+      addHtml(F("<TR><TD>Temperature:<TD>"));
+      addTaskSelect(F("plugin_145_temperature_task"), P145_PCONFIG_TEMP_TASK);
```
Yesterday, I made a change to the `addTaskSelect` function.
It now includes a `Not Set` option, which has an invalid TaskIndex_t value.
So right after this line, you chould check using `if (validTaskIndex(P145_PCONFIG_TEMP_MAX))` to see if you should continue where this taskindex is being used.

flashmark replied:
Done. Question: I see that the form is reloaded when another task is selected. Is this also possible if a checkbox is changed? I want to skip the compensation task/value settings when no compensation is selected.

TD-er replied:
I am already working on adding some extras on this `addTaskSelect` function to allow it to still reload and update the settings, but not (yet) saving it.
But it does have some side-effects, so I'm not 100% that will remain.

What do you want to do with it?
We can also add some JavaScript code to hide some elements.
For this I have added some functions which will specifically add a HTML id to wrappers so you can hide an entire row (or multiple rows) on change of some combo box. (a checkbox can also be done as a combo box)
I've done this for C018 and for the ESPEasySerial webforms functions.

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -0,0 +1,281 @@
+      if (P145_data != nullptr)
+      {
+        // we're checking a var from another task, so calculate that basevar
+        float temperature = UserVar[P145_PCONFIG_TEMP_TASK * VARS_PER_TASK + P145_PCONFIG_TEMP_VAL]; // in degrees C
```
Idem, check to see if `P145_PCONFIG_TEMP_TASK` is a valid task index.
Especially when trying to access elements in this `UserVar` array.

N.B. the same applies to `P145_PCONFIG_HUM_TASK`

TD-er commented:
Did another 'quick look' at the code and it for sure needs some attention before it gets merged.
Most of my remarks are just minor tweaks, but there are for sure some that really need to be changed.

For example the C-style casts, the cast to a flashstring and the missing check for `validTaskIndex` on the selected temp/hum task values.

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
Rework during review rework

flashmark commented:
Thanks for looking into my code. Indeed some remarks are serious and I tried to rework them properly. Please note that C++ experience is low, suggestions to improve are highly appreciated. 
With respect to the documentation. Also here it is sometimes difficult to grasp the complete structure, especially if the formatting language is new :-) I also doubt what the policy is here. As long as the plugin is not in any set should the document already be available? I will play with it tomorrow.

tonhuisman commented:
You can add it to `Collection F`, that has space available.

TD-er commented:
> I also doubt what the policy is here. As long as the plugin is not in any set should the document already be available?

The official policy is that we have the docs along with the code in the same repository, so you can be sure both are in sync.

Sadly the reality doesn't always agree with this policy.
I know I'm also guilty of not always having the documentation in sync with the code.

If your pull request covers a larger part of the code, you have to be sure not to keep it open too long or else merging will become an issue.
Also sometimes you just need to use the code for a while to see if things work out and then it could be that screenshots etc. become outdated. So it is OK to not have the docs finished before things get merged.
Just make sure users can operate the functionality. If it can be done without documentation or examples, then congratulations! You made the perfect and intuitive software imaginable.
For all other programmers... document as soon as possible :)

flashmark commented:
Question about form reload is the idea to show the temperature and humidity hask/value fields only when teh compensation checkbox is checked.  It will work as it is now, but why presenting fields that are not used by the software? However, that requires to redraw these fields when the checkbox state has changed. I am not a web-designer and have no idea how to implement this is ESPeasy.

tonhuisman commented:
> have no idea how to implement this is ESPeasy.

[This is a possible way](https://github.com/letscontrolit/ESPEasy/blob/mega/src/_P037_MQTTImport.ino#L155-L159) to do that

flashmark committed:
P145 fixes during review
P145 updated documentation

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark commented:
> > have no idea how to implement this is ESPeasy.

[This is a possible way](https://github.com/letscontrolit/ESPEasy/blob/mega/src/_P037_MQTTImport.ino#L155-L159) to do that 

Works as requested. Would be great to have this on checkboxes too. That enables showing more details when an option is switched on. Selection box does the job, but for me if feels a bit cluttering the screen.  A checkbox is a easier to operate.

tonhuisman commented:
> Would be great to have this on checkboxes too. That enables showing more details when an option is switched on. Selection box does the job, but for me if feels a bit cluttering the screen. A checkbox is a easier to operate.

Checkboxes are a bit awkward in html, and so far we never applied the same logic there. And also, 'merely' checking or un-checking a checkbox causing the page to reload may feel a bit daunting/unexpected to users, as often a mobile phone is used to operate the ESPEasy UI, where the view is already quite limited, selecting an explicit option from a list (a popup on your mobile browser) results in a better UX (not necessarily a better UI), IMHO.

flashmark committed:
P145 code cleanup,small  GUI changes, documentation updates

flashmark commented:
> > Would be great to have this on checkboxes too. That enables showing more details when an option is switched on. Selection box does the job, but for me if feels a bit cluttering the screen. A checkbox is a easier to operate.

Checkboxes are a bit awkward in html, and so far we never applied the same logic there. And also, 'merely' checking or un-checking a checkbox causing the page to reload may feel a bit daunting/unexpected to users, as often a mobile phone is used to operate the ESPEasy UI, where the view is already quite limited, selecting an explicit option from a list (a popup on your mobile browser) results in a better UX (not necessarily a better UI), IMHO. 

Clear and valid statement. I will stick to the selection boxes to keep the look and feel consistent within ESPeasy.

flashmark commented:
I am looking into some small improvements on the functionality. Nice to play with, but additional value for most users will be less. Are there any patters within ESPeasy to keep memory and other resource usage in balance with the offered feautres? I noticed for example de debugging flag and some memory foorprint reduction flag.

tonhuisman commented on src/_P145_MQxxx.ino:
```diff
@@ -175,10 +175,13 @@ boolean Plugin_145(byte function, struct EventStruct *event, String& string)
-      addFormNote(F("If this is enabled, the Temperature and Humidity values below need to be configured."));
-      //if (compensate)
+      //addFormCheckBox(F("Enable temp/humid compensation"), F("plugin_145_enable_compensation"), compensate);
+      addFormSelector_YesNo(F("Enable temp/humid compensation"), F("plugin_145_enable_compensation"), compensate, true);
```
To save some .bin size, please use short values for the id's, so instead of `plugin_145_enable_calibration` you could use `en_cal`, saving 24 * 2 = 48 bytes just in this single id... many of these smalls become 1 big saving 😃

flashmark commented:
> To save some .bin size, please use short values for the id's, so instead of `plugin_145_enable_calibration` you could use `en_cal`, saving 24 * 2 = 48 bytes just in this single id... many of these smalls become 1 big saving 😃 

I now remember I saw a macro to define a unique identifier, I assume it fits the requirements of being small but still unique. This is all legacy when I started with a copy of a very old plugin in the playground. So I appreciate all good recommendations.

flashmark committed:
P145 implement lowVcc feature, reduce memory usage, cleanup

TD-er commented on src/_P145_MQxxx.ino:
```diff
@@ -95,6 +95,19 @@
+#define P145_GUID_TEMP_T     (F("f08"))
+#define P145_GUID_TEMP_V     (F("f09"))
+#define P145_GUID_HUM_T      (F("f10"))
+#define P145_GUID_HUM_V      (F("f11"))
```
Better not use the F() macro in defines.
We always use it as 
```
#define P123_BLADIEBLA  "blaat"

F(P123_BLADIEBLA);
```

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
P145 Save some memory

flashmark committed:
P145 save some memory

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark commented:
> I need some help here as I am not a C++ expert. When I initialize teh fiels teh compiler complains about the hard coded "const struct P145_SENSORDEF sensorDefs[] PROGMEM" definition in P145_data_struct.cpp. I need a table with many sensors with each sensors characteristics and that data should be kept in PROGMEM. Only for the actual sensor the data is copied to RAM.
 

As old school C programmer C++ details fail. I want to define a struct that contains all kind of static (design time known) details about a particular sensor type. These shall be stored in PROGMEM as an array of this struct. Then I copy the struct of the selected sensor into RAM to be accessed through the fields of the stuct. This (seems to) work fine when having no initializers defined in teh struct. Once I define initializers the definition for the array in PROGMEM fails. Looks like teh compier suddenly interprets the stuct definition differently and tries to do something with constructors??? Any C++ expert that can shine a light on this?

TD-er commented:
Perhaps the enum type is causing issues here as it is not strictly defined what the `sizeof()` of this type is.
The enum is small enough to fit in a single byte, but typically the assumed type of an enum is `int`.

Here your struct:
https://github.com/letscontrolit/ESPEasy/blob/5660117e3959553804804d541deec15c5610030f/src/src/PluginStructs/P145_data_struct.h#L41-L56
I think you should not store it as an enum, but rather as an `int` or `uint32_t`.
Then to get it, you could add a function to the struct (functions are not part of the `sizeof()` of a class/struct)
For example:
```c++
P145_algorithm getAlg() const {
  return static_cast<P145_algorithm>(alg);
}
```

This way you know for sure the alignment of this struct and all the next ones is the same.
Maybe also a good idea to set this member `int alg;` as last member.
This may -in theory- be sub-optimal when aiming for the minimal size of a struct, but it will for sure make the `sizeof()` of this struct a multiple of 32 bits. This alignment is particularly important when storing stuff in PROGMEM, as flash-access is always in chunks of 32 bit.
Not sure if you can also give an attribute to progmem data to enforce some alignment on the progmem data.

TD-er commented:
One other thing that might be a bit funky here is the fact you're using C-strings.
For example `"USER"` will be of type `char name[5]` (zero terminated) and not `char name[8]`.
No idea how that may work out here.
So what you can do is create an enum for name and gas and then add a single `toString` function for those enums to convert them into a string.
This way you can store them as an uint8_t or something like that so you can be 100% sure about the size of the struct you try to save here.

These toString functions can also be part of the P145_SENSORDEF struct as you only need to get them and not set them.
Then your alg, name and gas can each be 8 bit ints, so best to add a 4th uint8_t type called "dummy" to pad it to a multiple of 4 bytes again.

flashmark commented:
About the structure issue: thanks for the hints. I will start plaing with this info, first finish tweaking the functionality of the pluging.

TD-er commented:
When adding those `toString` functions, you can let these return a flash string type.
The reason to return a flash string type instead of a `String` is that it reduces the number of `String()` constructor calls in the code.

For example:

```c++
#define ID_FOO 0
#define ID_BAR 1
String toString(uint8_t name) {
  switch (name) {
    case ID_FOO: return F("Foo");    
    case ID_BAR return F("Bar");
    default: break;
  }
  return F("");
}
```
This simple piece of code will actually compile into something using the `String(...)` constructor accepting a flash string for each return call.
Thus add some bytes on any return statement which is not needed.

In ESPEasy I have several of such functions where almost all values return a flash string and only a handful don't.
This would then force me to return a String type.
But instead I then have a single call to a function (only in the C++ file, not in the header, so not accessible from other code) to return a flash string for all other instances and if we need to create some String, we create a String type.
But for all other instances, we then only have 1 single return statement where this flash string is then converted to a String type.

flashmark commented:
I was playing with a version where I use the analog input to retrieve the value (no change) but also using a digital output to control a heater. I am quite confused how to use the Device[deviceCount].Type and how this is Type is reflected in IO handling. Somehow the plugin shall remember which pin is selected (CONFIG_PINx I assume). But how to combine an analog pin on ESP32 and the digital pin? Whenever I set Type= DEVICE_TYPE_DUAL I get a form with two digital pins to set up. But the analog input is setup with custom code.

TD-er commented:
You can also make it a custom type as I did with `_P148_POWR3xxD.ino`
Then you need to add pin select forms in the PLUGIN_WEBFORM_LOAD and make sure all will be saved.
In the mentioned plugin P148, I used `taskdevicepin1` ... `taskdevicepin3` as label, since I am also storing those pin nrs in `CONFIG_PIN1` ... `CONFIG_PIN3` as those are already being dealt with in the base call when you save a task.

The P098 PWM Motor plugin does have a combination of GPIO selection and some analog pin.

flashmark committed:
Removed strings from sensor definition table
Added optional heater control for MQ-7
Added new sensor definitions
Rework on algorithms

tonhuisman commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -11,14 +11,41 @@
+
+// Gas type identifiers used to condense the sensorDefs table to fixed format
+#define GAS_USER    0
+#define GAS_CO2     1
```
These #defines should have a `P145_` prefix, to avoid possible clashes with other (future) plugins.

tonhuisman commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -113,13 +186,39 @@ const struct P145_SENSORDEF sensorDefs[] PROGMEM =
+
+//Timeout values for Heater control
+#define HEATER_WARMUP_TIME (90*1000)
+#define HEATER_ON_TIME (60*1000)
```
Another case of missing `P145_` prefixes

tonhuisman commented on src/src/PluginStructs/P145_data_struct.cpp:
```diff
@@ -312,16 +415,27 @@ float P145_data_struct::getCorrectedPPM(float rSensor, float temperature, float
+      // No correction
+      c = 1.0f;
+  }
+  if (c <=0)  // In case no or wrong correction factors are spefified disable correction (multiplier = 1.0)
```
Float compare should also apply the decimal and f postfix, `0.0f`.
There's a typo in 'spefified'...

flashmark committed:
P145 updated documentation

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
P145 small documentation updates

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
P145 testing autocalibration

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark committed:
P145 plugin updates & refactoring

flashmark committed:
Merge branch 'letscontrolit:mega' into P145-Gases---MQxxx-(MQ135-CO2,-MQ3-Alcohol)

flashmark commented:
I used the custom pin setup and it works well. Except the pins are not added to the "pin state buffer" overview page. Is there a standard way to add them to the list of used pins?

flashmark commented:
I have been playing with the plugin and it seems pretty stable. Nu guarantee all sensors are giving sane values as I don't have calibration gasses. The MQ-135 trend indication is pretty ok, absolute values for these sensors seem to be difficult. Too mucxh influence from temperature & humidity. For me it can be merged as [testing]. I am going to spend time to other projects unless issues are reported.

flashmark committed:
P145 restructuring and some updates.

TD-er commented:
Just one more remark...
The P145_GUID_xxx defined now define a (flash) string.
However, they are not really descriptive, which is fine.
So why not define them as int values and then use the function `PCONFIG_LABEL` to generate a string?
Not sure how much we're saving here, if any, but at least it is slightly more consistent with other plugins.
These flash strings typically take some space to convert to string for each call, but on the other hand, I also have made function signature aliases for most of the functions used here to accept flash strings as an argument, so the gain in bin size might be very little here as these strings already are quite short.

flashmark commented:
I like to reuse patters and tried to use PCONFIG_LABEL. This has two drawbacks:
PCONFIG_LABEL expands to a function that returns a String. Not all functions liked to receive s string as input where the macro F() is accepted.
PCONFIG_LABEL seems to be intended to match the PCONFIG variables. It will not return a unique identifier string for the HTML labels.
String PCONFIG_LABEL(int n) {
  if (n < PLUGIN_CONFIGVAR_MAX) {
    return concat(F("pconf_"), n);
  }
  return F("error");
}

In this plugin parameters are stored in more locations. For example some bitfields are combined in one PCONFIG variable. Therefore the PCONFIG_LABEL does not work.
If there are other patterns available, please let me know. The current way of working is just invented after some review remarks.

TD-er commented:
OK, then leave it.
As I said, it will probably not save that much space since those strings are rather small already.
It is just that it 'feels' a bit wasteful to have flash strings for numbers.
But maybe I should just stop nagging and merge this otherwise good looking PR :)

flashmark commented:
I am used to the review process and can be strict on structure and reuse. Good to get known to the patterns used in ESPeasy.  I agree with both the attempt to be careful with memory and to try to use well known patterns. My knowledge about ESPeasy internals is too small to already promote generic functionality. However I wonder what the original intention with PCONFIG_LABEL was. I would expect a simple label generator unique per form. But it seems to be hard coupled to the PCONFIG concept itself, but the reason behind this connection is unknown to me.

As already clear I am not a well skilled C++ programmer. What still wonders me what mechanism allows me to define a hardcoded structured table in progmem and copy it into the same structure when using a C-style struct, but fails as soon as I add one initializer to the struct. Somehow the compiler changes this into a class-style struct definition and refuses to initialize the array.  But the used C-style construction now works fine.

flashmark commented:
Final comment on the sensors. They are a bit too unreliable and difficult to calibrate.  Using them for trend analysis is ok. Diving into the algorithms and learning the Plugin structure was however fun. But probabaly algorithms should be tuned or better sensor be bought :-) Thanks.

TD-er commented:
> As already clear I am not a well skilled C++ programmer. What still wonders me what mechanism allows me to define a hardcoded structured table in progmem and copy it into the same structure when using a C-style struct, but fails as soon as I add one initializer to the struct. Somehow the compiler changes this into a class-style struct definition and refuses to initialize the array. But the used C-style construction now works fine.

Not 100% sure I do understand what you mean here...
So my reply is purely based on 'speculation' on what you mean here :)

A struct and a class are the same in C++. The only difference is whether its members are `private` or `public` by default.
Still both need a constructor.
If you don't add one, the compiler will generate one without you realizing it.
When you create an object (which is a class or struct), there will first be some allocation of the size of the object.
However, this can be filled with garbage from previous data which was present at this memory location.
The constructor then calls the constructor of each member.
Some basic types like `int`, `float`, `char`, etc. do not have a constructor, which means their content can be just about anything.
For this you need to initialize these members in the constructor.
Since C++11, you can also give some 'hints' to the compiler by assigning some default value when declaring the members in the .h file. These will then be used by the compiler to generate the default constructor.

Since a member can also be a class or struct, these must also be constructed.
As long as an object is not constructed, its behavior is undefined.

C++ does have some extra tricks to make sure the members will actually be constructed and not just assigned some value.... maybe.
You can use the `{}` operator when declaring a member.
This will make sure the member is constructed and has some predictable value.
For the basic types, the value will be like `0.0f` for `float`, `false` for a `bool`, `0` for integer types, etc.
But now the really useful part.... arrays will also make sure all elements are constructed.
For example `String foo[10]{};` will make sure that all elements of `foo` are constructed.
That's something that has bitten me a few times, where `String foo[10];` may act strange when you try to access elements which were not actually set.

Another thing you need to keep in mind is that you should not call member functions of a class or struct from within the constructor of the same class/struct. Well at least not when using inheritance, but it is a good practice to never do this.
The problem is that during the construction of a class/struct, you can't be sure where the `this` pointer points to.
If you call a member function (e.g. calling the member function `toString();` ), the compiler turns such a call into `this->toString();`  But then you must be sure the `this` pointer is already present (and correct).

You should also keep in mind that some classes do allocate memory on the heap. Thus their member is then a pointer which is strictly runtime and should never to stored or retrieved. (unless you really, really know what you're doing)
For example the `String` class does allocate some memory on the heap, when the string is too long to fit in the default 11 bytes which are 'stack allocated'. N.B. this small array can also hold the pointer to heap-allocated memory.

Anyway, when you C-style copy the data into the same memory location as your class, you didn't call the constructor.
This may or may not work as it is simply undefined behavior.

flashmark commented:
I have ben on holiday, reading the comment just now. Issue is the usage of a hard coded table in PROGMEM.  This table consists of an array of structs. As soon as I define the struct with a default initializer teh compliter starts complaiuning about my table definition. Probabaly the compiler does not "understand" it should be a compile time defined table that will never change. Probabaly someting in the definition: const struct P145_SENSORDEF sensorDefs[] PROGMEM = {}

Error:
src/src/PluginStructs/P145_data_struct.cpp:212:1: error: could not convert '{0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, p145AlgNone, 0, 0}' from '<brace-enclosed initializer list>' to 'P145_SENSORDEF'
